README.md
# Highlight dodgy characters

Highlight all non-ascii characters (port of [Highlight Dodgy Characters](https://packagecontrol.io/packages/Highlight%20Dodgy%20Chars) from Sublime Text)

![Screenshot](./screenshot.png "Screenshot")

## Features

It highlights all the non-ascii characters, including dashes, fancy quotes and emojis. 

Inspired from [Highlight Bad Chars](https://github.com/WengerK/vscode-highlight-bad-chars/)

### Whitelisting

You can override the whitelist of characters that will not be highlighted in your workspace or user settings:
```
"highlight-dodgy-characters.whitelist": "´€£¡¿äàáâãåǎąăæçćĉčđďðèéêëěęĝģğĥìíîïıĵķĺļłľñńňöòóôõőøœŕřẞßśŝşšșťţþțüùúûűũųůŵýÿŷźžż"
```

### Examples

An example file that contains dodgy characters can be found [here](./src/test/examples.txt)

CHANGELOG.md
# Change Log
All notable changes to the "highlight-dodgy-characters" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

## [0.4.3] - 2019-03-17
### Added
- Setting for overriding the default whitelist
- Basic e2e test
### Changed
- Use TypeScript instead of JavaScript
- Improve README.md

src/test/extension.test.ts
import * as assert from 'assert';
import * as vscode from 'vscode'
import * as highlightChars from '../extension'

// Defines a Mocha test suite to group tests of similar kind together
suite("Extension Tests", function () {

    test("Should highlight bad characters in example file", () => {
        const editor = vscode.window.activeTextEditor;
        assert.ok(editor);
        const decorations = highlightChars.createDecorations(<vscode.TextEditor>editor);
        // the example file contains 18 dodgy characters
        assert.equal(decorations.length, 18);
        // make sure the first two decorators are where they should be
        assert.deepEqual(decorations[0].range.start, { _line: 0, _character: 18 });
        assert.deepEqual(decorations[1].range.start, { _line: 1, _character: 19 });
    });
});

extension.js
const vscode = require('vscode');

function activate(context) {
  const badCharDecorationType = vscode.window.createTextEditorDecorationType({
    cursor: 'crosshair',
    backgroundColor: 'rgba(255,0,0,0.3)',
    borderWidth: '1px',
    borderStyle: 'solid',
    borderColor: 'rgba(255,0,0,0.6)'
  });

  let whitelist = '\n´\u0009'; // allow newline, forward-tick and tabulator
  whitelist += '€£';
  // European characters (http://fasforward.com/list-of-european-special-characters/)
  whitelist +=
    '¡¿äàáâãåǎąăæçćĉčđďðèéêëěęĝģğĥìíîïıĵķĺļłľñńňöòóôõőøœŕřẞßśŝşšșťţþțüùúûűũųůŵýÿŷźžż';
  whitelist += '«»'; // guillemets

  // search for non-ascii characters that are not on the whitelist
  const charRegExp = `[^\x00-\x7F${whitelist}]`;

  let editor = vscode.window.activeTextEditor;
  if (editor) {
    triggerUpdateDecorations();
  }

  vscode.window.onDidChangeActiveTextEditor(
    editor => {
      editor = editor;
      if (editor) {
        triggerUpdateDecorations();
      }
    },
    null,
    context.subscriptions
  );

  vscode.workspace.onDidChangeTextDocument(
    event => {
      if (editor && event.document === editor.document) {
        triggerUpdateDecorations();
      }
    },
    null,
    context.subscriptions
  );

  var timeout = null;
  function triggerUpdateDecorations() {
    if (timeout) {
      clearTimeout(timeout);
    }
    timeout = setTimeout(updateDecorations, 500);
  }

  function updateDecorations() {
    if (!editor) {
      return;
    }
    let regEx = new RegExp(charRegExp, 'gi');
    const text = editor.document.getText();
    const badChars = [];
    let match;
    while ((match = regEx.exec(text))) {
      const startPos = editor.document.positionAt(match.index);
      const endPos = editor.document.positionAt(match.index + match[0].length);
      const decoration = {
        range: new vscode.Range(startPos, endPos),
        hoverMessage: 'Bad char "**' + match[0] + '**"'
      };
      badChars.push(decoration);
    }
    editor.setDecorations(badCharDecorationType, badChars);
  }
}
exports.activate = activate;
//# sourceMappingURL=extension.js.map


src/extension.ts
import * as vscode from 'vscode';
import { debounce } from 'lodash';

const configuration = vscode.workspace.getConfiguration('highlight-dodgy-characters');
const badCharDecorationType = vscode.window.createTextEditorDecorationType({
  cursor: 'crosshair',
  backgroundColor: 'rgba(255,0,0,0.3)',
  borderWidth: '1px',
  borderStyle: 'solid',
  borderColor: 'rgba(255,0,0,0.6)'
});

let whitelist = '\n\u0009'; // allow newline and tabulator
whitelist += configuration.whitelist;

// search for non-ascii characters that are not on the whitelist
const badCharMatcher = `[^\x00-\x7F${whitelist}]`;

export function activate(context: vscode.ExtensionContext) {
  // execute function on the leading edge of the debounce -> only defer subsequent calls
  const triggerUpdateDecorations = debounce(updateDecorations, 500, { leading: true });

  triggerUpdateDecorations();

  vscode.window.onDidChangeActiveTextEditor(
    editor => triggerUpdateDecorations(editor),
    null,
    context.subscriptions
  );

  vscode.workspace.onDidChangeTextDocument(
    event => {
      const editor = vscode.window.activeTextEditor;
      if (editor && editor.document === event.document) triggerUpdateDecorations(editor);
    },
    null,
    context.subscriptions
  );
}

function updateDecorations(editor: vscode.TextEditor | void) {
  if (!editor) editor = vscode.window.activeTextEditor;
  if (editor) {
    const decorations = createDecorations(editor);
    editor.setDecorations(badCharDecorationType, decorations);
  }
}

export function createDecorations(editor: vscode.TextEditor): vscode.DecorationOptions[] {
  const badCahrRegExp = new RegExp(badCharMatcher, 'gi');
  const text = editor.document.getText();
  const decorations = [];
  let match;
  while (match = badCahrRegExp.exec(text)) {
    const startPos = editor.document.positionAt(match.index);
    const endPos = editor.document.positionAt(match.index + match[0].length);
    const hexCharCode = match[0].charCodeAt(0).toString(16);
    const decoration = {
      range: new vscode.Range(startPos, endPos),
      hoverMessage: `Bad character: [${hexCharCode}](https://unicode.org/cldr/utility/character.jsp?a=${hexCharCode})`
    };
    decorations.push(decoration);
  }

  return decorations;
}